monai/apps/auto3dseg/ensemble_builder.py

```diff
@@ -0,0 +1,227 @@
+    """
+
+    def __init__(self):
+        self.algos = []
+        self.infer_files = []
+
+    def set_algos(self, infer_algos):
+        """
+        Register model in the ensemble
+        """
+        self.algos = deepcopy(infer_algos)
+
+    def get_algo(self, identifier):
+        """
+        Get a model by identifier.
+
+        Args:
+            identifier: the name of the bundleAlgo
+        """
+        for algo in self.algos:
+            if identifier == algo[ScriptEnsembleKeys.ID]:
+                return algo
+
+    def set_infer_files(self, dataroot: str, data_src_cfg_file: str):
+        """
+        Set the files to perform model inference.
+
+        Args:
+            dataroot: the path of the files
+            data_src_cfg_file: the data source file path
+        """
+        with open(data_src_cfg_file) as f:
+            datalist = json.load(f)
+
+        for d in datalist["testing"]:
+            self.infer_files.append({"image": os.path.join(dataroot, d["image"])})
+
+    @abstractmethod
+    def rank_algos(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def predict(self):
+        """
+        predict results after the models are ranked/weighted
+        """
+        raise NotImplementedError
+
+
+class EnsembleBestN(ScriptEnsemble):
+    """
+    Ensemble method that select N model out of all using the models' best_metric scores
+
+    Args:
+        n_best: number of models to pick for ensemble (N).
+    """
+
+    def __init__(self, n_best: int = 5):
+
+        super().__init__()
+        self.n_best = n_best
+
+    def sort_score(self):
+        """sort the best_metrics"""
+        scores = concat_val_to_np(self.algos, [ScriptEnsembleKeys.SCORE])
+        return np.argsort(scores).tolist()
+
+    def rank_algos(self):
+        """rank the model"""
+        ranks = self.sort_score()
+        if len(ranks) < self.n_best:
+            raise ValueError("Number of available algos is less than user-defined N")
+
+        # get the indices that the rank is larger than N
+        indices = [i for (i, r) in enumerate(ranks) if r >= self.n_best]
+
+        # remove the found indices
+        indices = sorted(indices, reverse=True)
+
+        self.ranked_algo = deepcopy(self.algos)
+        for idx in indices:
+            if idx < len(self.ranked_algo):
+                self.ranked_algo.pop(idx)
+
+    def predict(self, param: Optional[Dict[str, Any]] = None):
+        """predict the ensembled pred"""
+        if param is None:
+            param = {}
+
+        for i in range(len(self.infer_files)):
+            preds = []
+            infer_filename = self.infer_files[i]
+            for algo in self.ranked_algo:
+                infer_instance = algo[ScriptEnsembleKeys.ALGO]
+                param.update({"files": [infer_filename]})
+                pred = infer_instance.predict(param)
+                preds.append(pred[0])
+            output = sum(preds) / len(preds)
```

Comment:
this is an average, how about majority voting which is also common...

Reply:
I think we can support the `ensemble method` as a callable argument?

Thanks.

Reply:
Okay, I will make the ‘predict’ to a ‘__call__’ function.

Reply:
In the upcoming commit, I added the majority voting, made the BundleEnsemble a callable and some additional parameters to define the prediction.